spec/03-plugins/24-rate-limiting/05-integration_spec.lua

```diff
@@ -0,0 +1,177 @@
+local REDIS_PASSWORD = ""
+local REDIS_DATABASE = 1
+local REDIS_DATABASE_SECOND = 0
+
+local SLEEP_TIME = 1
+
+local function flush_redis(db)
+  local red = redis:new()
+  red:set_timeout(2000)
+  local ok, err = red:connect(REDIS_HOST, REDIS_PORT)
+  if not ok then
+    error("failed to connect to Redis: " .. err)
+  end
+
+  if REDIS_PASSWORD and REDIS_PASSWORD ~= "" then
+    local ok, err = red:auth(REDIS_PASSWORD)
+    if not ok then
+      error("failed to connect to Redis: " .. err)
+    end
+  end
+
+  local ok, err = red:select(db)
+  if not ok then
+    error("failed to change Redis database: " .. err)
+  end
+
+  red:flushall()
+  red:close()
+end
+
+describe("Plugin: rate-limiting (integration)", function()
+
+  local client
+
+  setup(function()
+    helpers.dao:drop_schema()
+    helpers.run_migrations()
+  end)
+
+  teardown(function()
+    flush_redis(REDIS_DATABASE)
+    flush_redis(REDIS_DATABASE_SECOND)
+    if client then client:close() end
+    helpers.stop_kong()
+  end)
+
+  describe("Redis conn select database", function()
+    -- Regression test for the following issue:
+    -- https://github.com/Kong/kong/issues/3292
+
+    setup(function()
+      flush_redis(REDIS_DATABASE)
+      flush_redis(REDIS_DATABASE_SECOND)
+      local api1 = assert(helpers.dao.apis:insert {
+        name         = "redistest1_com",
+        hosts        = { "redistest1.com" },
+        upstream_url = helpers.mock_upstream_url,
+      })
+      assert(helpers.dao.plugins:insert {
+        name   = "rate-limiting",
+        api_id = api1.id,
+        config = {
+          minute              = 1,
+          policy              = "redis",
+          redis_host          = REDIS_HOST,
+          redis_port          = REDIS_PORT,
+          redis_password      = REDIS_PASSWORD,
+          redis_database      = REDIS_DATABASE,
+          fault_tolerant      = false
+        },
+      })
+
+      local api2 = assert(helpers.dao.apis:insert {
+        name         = "redistest2_com",
+        hosts        = { "redistest2.com" },
+        upstream_url = helpers.mock_upstream_url,
+      })
+      assert(helpers.dao.plugins:insert {
+        name   = "rate-limiting",
+        api_id = api2.id,
+        config = {
+          minute              = 1,
+          policy              = "redis",
+          redis_host          = REDIS_HOST,
+          redis_port          = REDIS_PORT,
+          redis_password      = REDIS_PASSWORD,
+          redis_database      = REDIS_DATABASE_SECOND,
+          fault_tolerant      = false
+        }
+      })
+      assert(helpers.start_kong({
+        nginx_conf = "spec/fixtures/custom_nginx.template",
+      }))
+      client = helpers.proxy_client()
+    end)
+
+    it("redis conn select databases", function()
+      local red = redis:new()
+      red:set_timeout(2000)
+      local ok, err = red:connect(REDIS_HOST, REDIS_PORT)
+      if not ok then
+        error("failed to connect to Redis: " .. err)
```

Comment:
ditto: let's follow the Lua idioms and also make this a bit more readable (a few line jumps would be nice)